iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -233,18 +271,55 @@ public override async Task<Message> ReceiveAsync(TimeSpan timeout, CancellationT
                 }
             }
 
+            if (Logging.IsEnabled) Logging.Exit(this, timeoutHelper, timeoutHelper.RemainingTime(), $"{nameof(ReceiveAsync)}");
+
             return message;
         }
 
         async Task<bool> ReceiveMessageArrivalAsync(TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            bool hasMessage = false;
+            cancellationToken.ThrowIfCancellationRequested();
+            var disconnectToken = this.disconnectAwaitersCancellationSource.Token;
+            this.EnsureValidState();
+
+            using (CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, disconnectToken))
+            {
+                try
+                {
+                    hasMessage = await this.receivingSemaphore.WaitAsync(timeout, linkedCts.Token).ConfigureAwait(true);
```

Comment:
Can we do the same receivingSemaphore on AMQP instead of creating a whole new branch using TimeoutHelper?
That way the CancellationToken can be used without having the issues of maintaining two separate timeout modes while solving the issue of immediate cancellation.

Reply:
Can you explain more? What do you mean by two separate timeout modes here?